vcr/patch.py

```diff
@@ -18,13 +18,14 @@
     from botocore.awsrequest import AWSHTTPConnection, AWSHTTPSConnection
 except ImportError:
     try:
-        import botocore.vendored.requests.packages.urllib3.connectionpool as cpool
+        import botocore.vendored.requests  # noqa: F401
+
+        raise RuntimeError(
+            "vcrpy >=4.2.2 and botocore <1.11.0 are not compatible"
+            "; please upgrade botocore (or downgrade vcrpy)"
+        )
     except ImportError:  # pragma: no cover
         pass
```

Comment:
Can you also remove this try/except and the censored import? If the initial import failed raising the RuntimeError should be enough?

Additionally, should we use `raise  RuntimeError(...) from e`, e being the initial ImportError?

Reply:
> Can you also remove this try/except and the censored import? If the initial import failed raising the RuntimeError should be enough?

@pquentin I'm not sure I get the idea. If we can cut anything away here, I'll need your help to see it. Could you elaborate?


> Additionally, should we use `raise RuntimeError(...) from e`, e being the initial ImportError?

To my understand that `from e` is happening automatically and would need suppressing via `from None` if we wanted the opposite. To see it for yourself in practice, you could run these two programs:

Nested traceback output:
```python
try:
    raise ValueError('first')
except:
    raise ValueError('second')
```

Proof of set `__context__`:
```python
try:
    try:
        raise ValueError('first')
    except:
        raise ValueError('second')
except Exception as e:
    print(repr(e.__context__))
```

What do you think?

Reply:
@pquentin PS: quick update, I found one small difference when adding `from e1` to the above sample — the message in between the two tracebracks changes slightly:

- a) Default: `During handling of the above exception, another exception occurred:`
- b) `from e1`: `The above exception was the direct cause of the following exception:`

For myself, the difference is of little value. Do you see value in the difference?

Reply:
My message was not clear, because my phone autocorrected from vendored to censored. Anyway, I understand how we need the try/except because requests is an optional dependency.

Do we need `import botocore.vendored.requests` though?

> @pquentin PS: quick update, I found one small difference when adding from e1 to the above sample — the message in between the two tracebracks changes slightly: [...]

Yes, I think this is valuable, because in case a) you don't know if the exception handling failed or if the new exception was intentional.

Reply:
> 



> My message was not clear, because my phone autocorrected from vendored to censored. Anyway, I understand how we need the try/except because requests is an optional dependency.

@pquentin cool!

> Do we need `import botocore.vendored.requests` though?

I believe we do, because how else would we know that we are dealing with `botocore <1.11.0` without that import? In theory we could (a) use other means of detection or (b) not (be able to) throw the runtime error or (c) keep as is. What do you think?

> Yes, I think this is valuable, because in case a) you don't know if the exception handling failed or if the new exception was intentional.

I have added it now just for you :smile:

Reply:
Ah yes because botocore is optional so you need to handle the case where it is not here at all! Makes sense.

Maybe contextlib.suppress would have been slightly clearer and would have avoided the nocover pragma.

Anyway, thanks, you can consider this comment resolved! Thanks for taking the time to explain things.

Reply:
@pquentin while I like `contextlib.suppress` in general, to work here the `raise RuntimeError` would need to go into the `with` body and and I think that would make it a bit mind bending. but with that in mind I now have refactored it to pull the `raise RuntimeError` into a new `else` branch to hopefully be more obvious and more consistent to the rest of it. I hope that's good with you. Thanks for the review!